chrome/content/zotero/advancedSearch.js

```diff
@@ -60,13 +59,14 @@ var ZoteroAdvancedSearch = new function() {
 			id: "advanced-search",
 			dragAndDrop: true,
 			onActivate: this.onItemActivate.bind(this),
-			columns: getDefaultColumnsByDataKeys(['title', 'firstCreator']),
+			columns: Zotero.ItemTreeManager.getColumns([{dataKey: "title"}, {dataKey: "firstCreator"}], ["itemtree"]),
```

Comment:
Please, stop changing things around and let's discuss this. Was there any reason to change this original function? If not, can we just revert the original `getDefaultColumnsByDataKey()` functions?

Reply:
For posterity, this exchange looks weird, because we accidentally had pending comments from windingwind that were not submitted for days, due to Github's comment batching UI unclarity.

Regarding this function. We need to establish a distinction between live columns in the trees, and column definitions/specs in files like `itemTreeColumns.jsx`. `getDefaultColumnsByDataKeys()` was designed and is used in the code to get column specs to pass as initializing parameters to `ItemTree.init()`. The changes you made now make it retrieve live columns from trees, but there's no longer a way to retrieve column specs. I know I kind of encouraged it, although my initial comment was

> We should evaluate if we still need a function that only retrieves default columns and either add `getColumnsByDataKeys()` functions, or rename these.

so I kind of left the path open there.

We might want a function to retrieve live columns for plugins (although I'm not sure what they would do with that). But we also certainly need a function to just get good column spec for default Zotero item fields to be displayed in any item trees.

(As a side-note, in another branch I am working on I've added a column option `renderer`, which is a rendering function for that column. I didn't think about it before, but some of the functions in `itemTree.jsx` like `renderHasAttachmentCell()` should reasonably be replaced by better column spec, making the ItemTree more generic, and it also would likely come in handy for plugin developers, but we can think about it later)

So to summarize:

1. What would be the use case from a plugin perspective from a function that allows to retrieve live columns from a tree?
2. Unless I'm missing something, we should restore the original function (although we can think of a better name for it than `getDefaultColumnsByDataKeys()`.

Reply:
> What would be the use case from a plugin perspective from a function that allows to retrieve live columns from a tree?

One possible usage might be, say, there are at least 3 plugins that aim to provide citation count for papers and have some overlap in their functionalities. These plugins just register columns of their own to display their citation data, causing multiple columns that provide very similar data. You may ask why people install these plugins. Shouldn't it be enough to use just one of them? Well, each plugin has some extra features that people might rely on, that is also why there exists several plugins that provide similar major features. So if these plugins can get the columns dynamically, some may be able to skip registration of their column if some other columns exist, to avoid duplicate citation columns.

For my use case, I do not really rely on getting columns dynamically but simply don't want to look into the source code to find out how to import and call the new function that can return a group of new default keys (like the cite item tree, which I could possibly use when we have it).

And I would need the ability to filter columns by multiple properties (like multiple dataKeys), to get a filtered columns group as the default columns of my item tree view.

> Unless I'm missing something, we should restore the original function (although we can think of a better name for it than getDefaultColumnsByDataKeys().

Why I dislike `getDefaultColumnsByDataKeys` is just because it is conceptually a sub-collection of `getColumns`. We may but that seems not to be very necessary, as `getColumns` can do what it can do and also what it can't, with relatively simple syntax.

> Regarding this function. We need to establish a distinction between live columns in the trees,

The second arg of `getColumns` can play a role like a filter to the data source of columns, the custom ones, the main ones, or other sources we add in the future.

Reply:
> The second arg of `getColumns` can play a role like a filter to the data source of columns, the custom ones, the main ones, or other sources we add in the future.

There is only one default spec for the author, attachment, notes, etc. columns. If I am creating a new ItemTree that needs to display the item title, creator and some other custom columns, I don't want to have to manually re-specify those, nor do I want to get them from any live item tree. I just need to get the default spec for the columns. There is and should be a single source of truth for that - `COLUMNS` in `itemTreeColumns.jsx`. These don't rely on any specific instantiation of an ItemTree, they just define how certain Zotero Item properties should be displayed in an ItemTree. I don't really see an instance of ever wanting to create an item tree, where you only get columns that e.g. persist width, or have staticWidth, or similar either. When creating a dialog with an ItemTree, you'll know which columns you need beforehand.

> One possible usage might be, say, there are at least 3 plugins that aim to provide citation count for papers and have some overlap in their functionalities. These plugins just register columns of their own to display their citation data, causing multiple columns that provide very similar data. You may ask why people install these plugins. Shouldn't it be enough to use just one of them? Well, each plugin has some extra features that people might rely on, that is also why there exists several plugins that provide similar major features. So if these plugins can get the columns dynamically, some may be able to skip registration of their column if some other columns exist, to avoid duplicate citation columns.

We can have `ItemTreeManager.getColumns()` for this.

> For my use case, I do not really rely on getting columns dynamically but simply don't want to look into the source code to find out how to import and call the new function that can return a group of new default keys (like the cite item tree, which I could possibly use when we have it).

If you're writing an UI for Zotero that uses an ItemTree, you will have to look at the source code in many places, including figuring out how to instantiate the columns that you want, so this is a moot point. The document citations dialog is not an extension of a tree, it uses a tree to create an UI, so it's not something to "use" as a development API per se.

Reply:
The current `ItemTreeManager.getColumns()` is doing multiple things right now. We need a proper `getCustomColumns()` function to be used in the `ItemTree.getColumns()` function, a separate function to get column specs, like the removed `getDefaultColumnByDataKey()`, and if we want to have something to get from live ItemTrees, like `main` tree, and `feeds` tree, we need a `ItemTreeManager.getColumns()` function that gets columns from live trees, but we need additional code to do that, mainly, something that registers all live ItemTrees so that they can be accessed and live columns retrieved.

Reply:
Well, I'd like to restore the original `getDefaultColumnByDataKey`, but before that, I still have some points not clear:

> I don't want to have to manually re-specify those, nor do I want to get them from any live item tree.

That might be `Zotero.ItemTreeManager.getColumns(null, ['itemtree'])`, which exactly returns the `COLUMNS` in itemTreeColumns.jsx. It's not from a live tree (although you can get columns of the live tree using `getColumns`), but a static one-truth source. Furthermore, if you want to do some filter stuff, then you get what you see in this line. 

We could also make a different default data source (like by default it can be 'itemtree', making this call be `Zotero.ItemTreeManager.getColumns()`.

We can also make new methods for getting default or custom columns, working as a syntactic sugar based on `getColumns`.

I write the current version because at some point you mentioned 'do not treat default columns specially', so I thought they should have the same level as the custom columns and be managed in one place. And based on that and the discussions above, I am a little confused, as seems like we are actually treating it specially.

Reply:
> Well, each plugin has some extra features that people might rely on, that is also why there exists several plugins that provide similar major features. So if these plugins can get the columns dynamically, some may be able to skip registration of their column if some other columns exist, to avoid duplicate citation columns.

I haven't followed the discussion above, but this example doesn't make any sense to me. If a plugin provides a feature, it should provide the feature. It could have an option to disable it, but it would be bizarre and confusing for a plugin to check whether some other plugin is available to decide whether to provide its own version of the feature.

Reply:
Yeah, it does not make sense, but that's actually happening. Take the pdf-preview plugin as an example, I have to detect if BBT changes the item box layout and based on that, run the corresponding code to make the preview pane looks fine.

Reply:
That's about another API that I will want to do later, but this kind of behavior widely exists. Every plugin does things in its way, partly because there is no API. The new plugins have to give way to those old plugins, to be compatible to the old ones.

Reply:
I'll admit I hadn't thought it through originally when we started making changes there, but I once again want to stress the difference between column definition and custom columns. If a plugin wants to augment the advanced search item tree with a `year` column, they should be able to. If a plugin wants to augment the library item tree with a `feed` column, they should be able to. I.e. `COLUMNS` in `itemTreeColumns.jsx` are **definitions** of columns that loosely correspond to Zotero Item fields, not an exhaustive list of columns on any particular tree. As such, the column registration code should not be using that list to prevent registration with particular data keys. Moreover, if one plugin wants to register a `custom` column on the main item tree, and another plugin wants to register a `custom` column on the advanced search tree, that should also work.

Additionally, the reason for existence of the previous `getDefaultColumnsByDataKey()` function was to retrieve column definitions when instantiating a new tree. We still need that function. It could more precisely be named `getColumnDefinitionsForDataKey()`. There really isn't any need for any additional filtering functions there and I can't think of any custom ItemTree you could make where you would need anything else than data keys, it was really just a helper function I added, since I was writing code like `COLUMNS.filter().map(c => Object.assign({}, c))` and wanted it to be clearer and make sure that other retrievals didn't accidentally modify the column definitions, by not cloning the definition object.

The function that retrieves custom registered columns is separate. It's never meant to be used when instantiating the tree, and when used within the tree code to get a final list of columns, it should never return non-custom columns, since any initial columns to the tree are passed in as a prop.

There can be another function that returns ALL live columns from a given item tree, if we think we need that. That function would not return the correct results unless the tree in question was actually instantiated, since it needs to know what `columns` prop got passed into a tree, which only happens when the tree is instantiated. I don't see a good reason for such a function to exist though, since if we want to offer something for plugin developers so that they can avoid collisions with other plugins, they only need to retrieve custom columns. Although I could agree that maybe that's not something we should be encouraging plugin developers to think by providing such an API.

Reply:
We are not. In fact this PR prevents duplicate column ids. It's probably a good idea to just namespace them instead (although that strays strictly into a not-dogfooding plugin APIs territory).

Reply:
Giving them a prefix when `pluginID` is provided would be a good way to namespace it. Should we keep the `getColumns` as it is, or make it something like `getCustomColumnsByKey`, once we have the previous `getxxx` in `itemTreeColumns.jsx` back?

I'm also not sure why we keep `itemTreeColumns` as a JSX, for it seems to have nothing to do with react.

Reply:
> Giving them a prefix when `pluginID` is provided

If we're going to be namespacing the columns, `pluginID` should be required.

> Should we keep the getColumns as it is, or make it something like getCustomColumnsByKey, once we have the previous getxxx in itemTreeColumns.jsx back?

I don't see any use of the `getCustomColumns()` function with extra filtering for use cases by Zotero. We just need a function to retrieve custom columns registered for a particular tree id. And I don't generally see plugin developers needing to use it at all, and if they do, rarely will they need to get a custom column by specific property, but if they need to, they can still get all columns and filter in their own code. So I think something more basic would suffice.

> I'm also not sure why we keep itemTreeColumns as a JSX, for it seems to have nothing to do with react.

Some column icons are React elements specified with the jsx syntax (and need to be preprocessed).

Reply:
Namespaced.

Reply:
I didn't see your comment until after the commit that namespace dataKey.

> `pluginID` should be required.

The current way is to give the `dataKey` a prefix of pluginID if provided, otherwise, keep the dataKey as it is. If we are going to use these APIs, we may not be able to have a plugin id. But for plugins, we encourage them to do that (and they will because they do not need to manually remove the columns when unloaded)